Source/SwiftLintFramework/Rules/Lint/DuplicateConditionsRule.swift

```diff
@@ -0,0 +1,227 @@
+                  bar()
+                }
+            """),
+            Example("""
+                if ↓#available(macOS 10.15, *) {
+                  foo()
+                } else if ↓#available(macOS 10.15, *) {
+                  bar()
+                }
+            """),
+            Example("""
+                if ↓case .p = x {
+                  foo()
+                } else if ↓case .p = x {
+                  bar()
+                }
+            """)
+        ]
+    )
+
+    func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+}
+
+private extension DuplicateConditionsRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override func visitPost(_ node: IfExprSyntax) {
+            if  node.parent?.is(IfExprSyntax.self) == true {
+                // We can skip these cases - they will be picked up when we visit the top level `if`
+                return
+            }
+
+            var maybeCurr: IfExprSyntax? = node
+            var statementChain: [IfExprSyntax] = []
+            while let curr = maybeCurr {
+                statementChain.append(curr)
+                maybeCurr = curr.elseBody?.as(IfExprSyntax.self)
+            }
+
+            let positionsByConditions = statementChain
+                .reduce(into: [Set<String>: [AbsolutePosition]]()) { acc, elt in
+                    let conditions = elt.conditions.map(extract)
+                    let location = elt.conditions.positionAfterSkippingLeadingTrivia
+                    acc[Set(conditions), default: []].append(location)
+                }
+
+            addViolations(positionsByConditions.values)
+        }
+
+        override func visitPost(_ node: SwitchCaseListSyntax) {
+            let switchCases = node.compactMap { $0.as(SwitchCaseSyntax.self) }
+
+            let positionsByCondition = switchCases
+                .reduce(into: [String: [AbsolutePosition]]()) { acc, elt in
+                    // Defaults don't have a condition to worry about
+                    guard case let .case(caseLabel) = elt.label else { return }
+                    for caseItem in caseLabel.caseItems {
+                        let pattern = caseItem
+                            .pattern
+                            .debugDescription(includeChildren: true, includeTrivia: false)
+                        let whereClause = caseItem
+                            .whereClause?
+                            .debugDescription(includeChildren: true, includeTrivia: false)
+                            ?? ""
+                        let location = caseItem.positionAfterSkippingLeadingTrivia
+                        acc[pattern + whereClause, default: []].append(location)
+                    }
+                }
+
+            addViolations(positionsByCondition.values)
+        }
+
+        private func extract(_ node: ConditionElementSyntax) -> String {
```

Comment:
I think this method now boils down to only the single expression/statement `node.debugDescription(includeChildren: true, includeTrivia: false)`, i.e. the switch statement isn't needed anymore.

Reply:
Yup sounds good; I think they're very mildly different, e.g., 
```
ConditionElementSyntax children=1
  0: IdentifierExprSyntax children=1
    0: identifier("z")
```
vs
```
IdentifierExprSyntax children=1
  0: identifier("z")
```

But I'd agree, I don't think the distinction matters in this context. Thanks!

Reply:
Have just been playing with this, and I think it's correct as is; it's subtle, though.

The use of `node` in `node.debugDescription` is referring to the inner `node`, from the `case let` in the switch statements; e.g., of type `SequenceExprSyntax`, or `ExprListSyntax`, etc. If we get rid of the switch statements, we are throwing away the 'unwrapping', and looking just at the outer `ConditionElementSyntax` type.

This is an issue when we have multiple conditions on the same line, because the outer `ConditionElementSyntax` keeps information such as trailing commas. E.g., comparing strings such as

```
ConditionElementSyntax children=2
  0: SequenceExprSyntax children=1
    0: ExprListSyntax children=3
      0: IdentifierExprSyntax children=1
        0: identifier("y")
      1: BinaryOperatorExprSyntax children=1
        0: binaryOperator("==")
      2: StringLiteralExprSyntax children=3
        0: stringQuote
        1: StringLiteralSegmentsSyntax children=1
          0: StringSegmentSyntax children=1
            0: stringSegment("s")
        2: stringQuote
  1: comma
```

Behaves differently to comparing strings such as
```
SequenceExprSyntax children=1
  0: ExprListSyntax children=3
    0: IdentifierExprSyntax children=1
      0: identifier("y")
    1: BinaryOperatorExprSyntax children=1
      0: binaryOperator("==")
    2: StringLiteralExprSyntax children=3
      0: stringQuote
      1: StringLiteralSegmentsSyntax children=1
        0: StringSegmentSyntax children=1
          0: stringSegment("s")
      2: stringQuote
```

I'll rename the inner `node` to something else to avoid the name overload, but I think the behaviour as written is correct

Reply:
You are right. The comma might interfere. So let me correct my proposal to
```swift
node.with(\.trailingComma, nil).debugDescription(includeChildren: true, includeTrivia: false)
```
Honestly, the only thing that bothers me in the current switch is the duplication of the `.debugDescription(includeChildren: true, includeTrivia: false)` calls.

Reply:
My bad; I wasn't taking `.condition`. You're right and we can do away with the switch. Done